### PR TITLE
Fix install target

### PR DESCRIPTION
**Pull request by jacereda**
Only tested on Darwin with Nixpkgs.

### CODE REVIEW

File: dev/src/platform.cpp [hunk 1]
```diff
@@ -197,9 +197,7 @@ bool InitPlatform(const char *exefilepath, const char *auxfilepath, bool from_bu
             write_dir = exepath;
         }
         data_dirs.push_back(string(exepath));
-        #ifdef __linux__
-            data_dirs.push_back("/usr/share/lobster/");
-        #endif
+        data_dirs.push_back(DATADIR);
```
Comment (aardappel) [inline dev/src/platform.cpp]: Please surround with `#ifdef DATADIR`, this is built on many platforms with many build systems, not just CMake

File: dev/CMakeLists.txt [hunk 2]
```diff
@@ -109,3 +109,15 @@ target_link_libraries(${EXE_NAME}
   ${SDL_LIBRARIES}
   ${ADDITIONAL_LIBRARIES}
   ${OPENGL_LIBRARIES})
+
+if(UNIX)
+   include(GNUInstallDirs)
+   set(BINDIR ${CMAKE_INSTALL_FULL_BINDIR})
+   set(DATADIR ${CMAKE_INSTALL_FULL_DATADIR})
```
Comment (aardappel) [inline dev/CMakeLists.txt]: `DATADIR` is too to clash as a preprocessor symbol, maybe `PLATFORM_DATADIR` ?

File: dev/CMakeLists.txt [hunk 3]
```diff
@@ -109,3 +109,15 @@ target_link_libraries(${EXE_NAME}
...
+else()
+   set(BINDIR .)
+   set(DATADIR .)
+endif()
+set_source_files_properties(src/platform.cpp PROPERTIES COMPILE_FLAGS -DDATADIR="\\\"${DATADIR}/lobster-data/\\\"")
```
Comment (aardappel) [inline dev/CMakeLists.txt]: Why `lobster-data` instead of `lobster` ? The location already indicates it is data.

File: dev/CMakeLists.txt [hunk 4]
```diff
@@ -109,3 +109,15 @@ target_link_libraries(${EXE_NAME}
   ${SDL_LIBRARIES}
   ${ADDITIONAL_LIBRARIES}
   ${OPENGL_LIBRARIES})
+
+if(UNIX)
+   include(GNUInstallDirs)
+   set(BINDIR ${CMAKE_INSTALL_FULL_BINDIR})
+   set(DATADIR ${CMAKE_INSTALL_FULL_DATADIR})
+else()
```
Comment (aardappel) [inline dev/CMakeLists.txt]: remove `else` case? We don't want to add the current dir as `DATADIR`.

Comment (jacereda) [reply dev/CMakeLists.txt]: I'm thinking of Windows here, so you can do something like this (untested)

```
cmake -DCMAKE_INSTALL_PREFIX=distribution .
make install
```

And end up with a `distribution` directory containing the required bits.

File: dev/CMakeLists.txt (hunk 3 above)
Comment (jacereda) [reply dev/CMakeLists.txt]: On non-unix platforms, the installed hierarchy will contain something like:
```
lobster
lobster-data/
```

I can't name both the directory and the executable `lobster`.

File: dev/CMakeLists.txt (hunk 2 above)
Comment (jacereda) [reply dev/CMakeLists.txt]: Right, done.

File: dev/src/platform.cpp (hunk 1 above)
Comment (jacereda) [reply dev/src/platform.cpp]: Done.

File: dev/CMakeLists.txt (hunk 3 above)
Comment (jacereda) [reply dev/CMakeLists.txt]: Could be just `data` I guess, but then there will be a `data/data` directory.

File: dev/CMakeLists.txt (hunk 4 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: On non-unix platforms, we're currently not using CMake, and we'd like to keep the directory structure as-is. E.g. both VS and Xcode produce an executable in the `lobster` dir. They don't need this additional data dir.

Comment (jacereda) [reply dev/CMakeLists.txt]: Should I just remove the whole `if` and unconditionally use `GNUInstallDirs`?

File: dev/CMakeLists.txt (hunk 3 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: We're not doing this on on non-unix platforms. Lets make sure these modifications you're making are only for platforms that desire to have data in `/usr/share` or similar. For others we're keeping existing layout.

On a side-note, I'd like to rename the `lobster` dir, since it is way too confusing in documentation and scripts. Not sure what to call it though.. any ideas?

File: dev/CMakeLists.txt (hunk 4 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: No, again, don't want this to affect other platforms.

Comment (jacereda) [reply dev/CMakeLists.txt]: Which non-unix platforms are built via CMake and what should a `make install` look like on those?

File: dev/CMakeLists.txt (hunk 3 above)
Comment (jacereda) [reply dev/CMakeLists.txt]: What about getting rid of the `lobster` directory and moving everything up one level?

File: dev/CMakeLists.txt (hunk 4 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: None.
To be clear, we have these 6 platforms:

Windows: VS
OS X & iOS: XCode
Android: Android SDK/NDK
WebAssembly: Makefile
Linux: CMake

Theoretically other unices might build with the CMake file, but I don't think anyone has ever tried.

I may want to switch Windows to CMake at some point, but when I do, it shouldn't be trying to install files anywhere, just use the files in-place. So make install in system dir code only take effect on Linux for now.

File: dev/CMakeLists.txt (hunk 3 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: I was going to move `docs` up, but it still makes sense to have the other files in one dir?

The original idea was the `lobster` dir is everything a user of the language needs, including the exe in that dir (which is Windows centric, sorry). Not sure if I want to dump all of that in the root.

Maybe you're right though.. will have to think about it.

Comment (aardappel) [reply dev/CMakeLists.txt]: Well, you got your wish.. moved everything in `lobster/*` to the root. Easier this way.

File: dev/CMakeLists.txt (hunk 4 above)
Comment (jacereda) [reply dev/CMakeLists.txt]: What about this solution? I haven't tested, but on Unix installing to the sources dir seems to work. Files won't be copied (src and dst are the same), so only the lobster executable will be copied on `make install`.

Comment (aardappel) [reply dev/CMakeLists.txt]: Again, on platforms that are non a Linux-alike (which is most of them), the existing behavior needs to be preserved. This changes the Windows exe for example to not be in `bin` anymore, but in some install dir? There is no "installing" on this platform. Please make your changes affect Linux only. I thought the whole point of this PR was just to allow us to do Linux packaging? So lets not use `UNIX` as it also includes all Apple platforms.

Comment (jacereda) [reply dev/CMakeLists.txt]: - You said only Linux uses CMake.
- Even if you switched the Windows build to CMake, this behaviour is more reasonable. It should generate the lobster binary in the checkout directory. If you no longer want that behaviour, it should be easy to put it in a `bin` directory. 
- How can an `install` target for non-Unix be a bad thing? It should allow an easy isolation of required files for distribution (everything except the `dev` folder). If you just want to work in an install-less setup, just don't do `make install` and there you are.

**Comment by jacereda** [comment]
Looks like the Linux CI is failing because there's a `lobster` directory present, could that be due to a previous build?

File: dev/CMakeLists.txt (hunk 4 above)
Comment (aardappel) [reply dev/CMakeLists.txt]: It currently does.. even so, OS-specific stuff should be in the proper conditionals. Even on Linux, it should end up in the `bin` dir during development, when not explicitly using `make install`. And there's nothing "more reasonable" given such a wide diversity of OS standards. The concept of "install" in this sense doesn't make sense for Windows (unless it generated an installer).

Comment (jacereda) [reply dev/CMakeLists.txt]: It's already generating the unix executables under `bin`, check the last commit. And I have to disagree, `install` makes sense on Windows. Just tell it where to generate a distributable folder with `-DCMAKE_INSTALL_PREFIX=distributable`.

Comment (aardappel) [reply dev/CMakeLists.txt]: Ok, I guess we disagree. If you insist, it can go in, I can always change it once I get to use it on Windows.

File: dev/CMakeLists.txt [hunk 5]
```diff
@@ -4,26 +4,20 @@ project(Lobster)
...
+set(VIDEO_VULKAN OFF CACHE INTERNAL "")
+add_subdirectory(external/SDL EXCLUDE_FROM_ALL)
 
 file(GLOB LOBSTER_SRCS
     "src/*.h"
     "src/lobster/*.h"
     "src/*.cpp"
 )
 
-if(LOBSTER_TOCPP)
```
Comment (aardappel) [inline dev/CMakeLists.txt]: Why was this removed?

Comment (jacereda) [reply dev/CMakeLists.txt]: I moved it down so it's after the CMAKE_RUNTIME_OUTPUT_DIRECTORY setting, otherwise that setting won't take effect.

Comment (jacereda) [reply dev/CMakeLists.txt]: TBH I didn't verify the `LOBSTER_TOCPP` case, but I assumed the `set(EXE_NAME` required `CMAKE_RUNTIME_OUTPUT_DIRECTORY`.

Comment (aardappel) [reply dev/CMakeLists.txt]: Ah ok, didn't spot that.

**Comment by aardappel** [comment]
Thanks!

**Comment by aardappel** [comment]
So this did break the `bin` dir when using an "out of source" build, so had to revert it back to https://github.com/aardappel/lobster/commit/649d922d53541ab7b3880d19850414a3b0fa6dcc
